### PR TITLE
[WIP] feature: support delete on sharded table

### PR DESCRIPTION
**Pull request by dk-lockdown**
ref: https://github.com/cectc/dbpack/issues/44
### Ⅰ. Describe what this PR did

Todo: Need to handle distributed transactions across databases

### Ⅲ. Why don't you add test cases (unit test/integration test)? 
there are unit test cases and integration test cases

### Ⅳ. Describe how to verify it
check ci result

### CODE REVIEW

**Comment by tanryberdi** [comment]
@dk-lockdown LGTM, one suggestion, this PR is too big for 1 commit

**Comment by dk-lockdown** [comment]
> @dk-lockdown LGTM, one suggestion, this PR is too big for 1 commit

yes, the delete operation is new feature, and refactored query and insert operations.

File: pkg/optimize/optimize_insert.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+		tableMeta, err = meta.GetTableMetaCache().GetTableMeta(ctx, sqlDB, tables[0])
+		if err != nil {
+			continue
+		} else {
+			break
+		}
+	}
+	pk := tableMeta.GetPKName()
+	index := findPkIndex(stmt, pk)
+	// todo if index == -1, automatic generate a pk
```
Comment (bohehe) [inline pkg/optimize/optimize_insert.go]:
```suggestion
	// todo if index == -1, automatically generate a primary key
```

File: pkg/optimize/optimize_insert.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	pkValue := getPkValue(ctx, stmt, index, args)
```
Comment (bohehe) [inline pkg/optimize/optimize_insert.go:59]: Need to return error if `index==-1`, or `stmt.Lists[0][pkIndex]` will panic.

File: pkg/plan/delete.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		err                    error
+	)
+	tx, _, err = p.Executor.Begin(ctx)
+	if err != nil {
+		return nil, 0, errors.WithStack(err)
+	}
+	for _, table := range p.Tables {
+		sb.Reset()
+		if err = p.generate(&sb, table); err != nil {
+			return nil, 0, errors.Wrap(err, "failed to generate sql")
```
Comment (bohehe) [inline pkg/plan/delete.go]:
```suggestion
			return nil, 0, errors.Wrap(err, "failed to generate sql for delete")
```

File: pkg/plan/delete.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+		ctx.WriteKeyWord(" WHERE ")
+		if err := p.Stmt.Where.Restore(ctx); err != nil {
+			return errors.Wrap(err, "An error occurred while restore DeleteStmt.Where")
+		}
+	}
+
+	if p.Stmt.Order != nil {
+		ctx.WritePlain(" ")
+		if err := p.Stmt.Order.Restore(ctx); err != nil {
+			return errors.Wrap(err, "An error occurred while restore DeleteStmt.Order")
```
Comment (bohehe) [inline pkg/plan/delete.go]:
```suggestion
			return errors.Wrap(err, "An error occurred while restoring DeleteStmt.Order")
```

File: pkg/plan/delete.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+func (p *DeleteOnSingleDBPlan) generate(sb *strings.Builder, table string) error {
+	ctx := format.NewRestoreCtx(format.DefaultRestoreFlags, sb)
+	ctx.WriteKeyWord("DELETE ")
+	ctx.WriteKeyWord("FROM ")
+	// todo add xid hint for distributed transaction
+	ctx.WritePlain(table)
+	if p.Stmt.Where != nil {
+		ctx.WriteKeyWord(" WHERE ")
+		if err := p.Stmt.Where.Restore(ctx); err != nil {
+			return errors.Wrap(err, "An error occurred while restore DeleteStmt.Where")
```
Comment (bohehe) [inline pkg/plan/delete.go]:
```suggestion
			return errors.Wrap(err, "An error occurred while restoring DeleteStmt.Where")
```

File: pkg/plan/delete.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+	}
+
+	if p.Stmt.Order != nil {
+		ctx.WritePlain(" ")
+		if err := p.Stmt.Order.Restore(ctx); err != nil {
+			return errors.Wrap(err, "An error occurred while restore DeleteStmt.Order")
+		}
+	}
+
+	//if p.Stmt.Limit != nil {
```
Comment (bohehe) [inline pkg/plan/delete.go:94]: why not uncomment these code?

File: pkg/plan/insert.go [hunk 7]
```diff
@@ -42,36 +44,56 @@ func (p *InsertPlan) Execute(ctx context.Context) (proto.Result, uint16, error)
...
+			if i != 0 {
+				ctx.WritePlain(",")
+			}
+			ctx.WritePlain("(")
+			for j, v := range row {
+				if j != 0 {
+					ctx.WritePlain(",")
+				}
+				if err := v.Restore(ctx); err != nil {
+					return errors.Wrapf(err, "An error occurred while restore InsertStmt.Lists[%d][%d]", i, j)
```
Comment (bohehe) [inline pkg/plan/insert.go]:
```suggestion
					return errors.Wrapf(err, "An error occurred while restoring InsertStmt.Lists[%d][%d]", i, j)
```

**Review by bohehe — CHANGES_REQUESTED** [review]

File: pkg/plan/delete.go (hunk 6 above)
Comment (dk-lockdown) [reply pkg/plan/delete.go:94]: The delete operation may limit the number of deleted items, but after sharding, adding a deletion limit to each shard will result in an inaccurate number of items deleted.

**Review by bohehe — APPROVED** [review]
LGTM

**Comment by fatelei** [comment]
LGTM

**Pull request merged**